src/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/clusterclient/clusterinteraction/ClusterDetailFragment.kt

```diff
@@ -19,35 +44,181 @@ import kotlinx.coroutines.cancel
  * ClusterDetailFragment allows user to pick cluster, command, specify parameters and see
  * the callback result.
  */
-class ClusterDetailFragment : Fragment(){
+class ClusterDetailFragment : Fragment() {
   private val deviceController: ChipDeviceController
     get() = ChipClient.getDeviceController(requireContext())
 
   private val scope = CoroutineScope(Dispatchers.Main + Job())
+  private lateinit var clusterMap: HashMap<String, ClusterInfo>
+  private lateinit var selectedClusterInfo: ClusterInfo
+  private lateinit var selectedCluster: ChipClusters.BaseChipCluster
+  private lateinit var selectedCommandCallback: DelegatedClusterCallback
+  private var commandArguments = HashMap<String, Any>()
+  private lateinit var selectedCommandInfo: CommandInfo
+  private var devicePtr = 0L
+  private var endpointId = 1
 
   override fun onCreateView(
     inflater: LayoutInflater,
     container: ViewGroup?,
     savedInstanceState: Bundle?
   ): View {
+    clusterMap =
+      checkNotNull(requireArguments().getSerializable(CLUSTER_MAP_INFO)) as HashMap<String, ClusterInfo>
+    devicePtr = checkNotNull(requireArguments().getLong(DEVICE_PTR))
     return inflater.inflate(R.layout.cluster_detail_fragment, container, false).apply {
       deviceController.setCompletionListener(GenericChipDeviceListener())
+      commandAutoComplete.visibility = View.GONE
+      clusterAutoCompleteSetup(clusterAutoComplete, commandAutoComplete, parameterList)
+      commandAutoCompleteSetup(commandAutoComplete, inflater, parameterList, callbackList)
+      invokeCommand.setOnClickListener {
+        parameterList.forEach {
+          val type =
+            selectedCommandInfo.commandParameters[it.parameterName.text.toString()]!!.type!!
+          val data = castCorrectType(type, it.parameterData.text.toString())!!
+          commandArguments[it.parameterName.text.toString()] = data
+        }
+
+        selectedCommandInfo.getCommandFunction()
+          .invokeCommand(selectedCluster, selectedCommandCallback, commandArguments)
+      }
+    }
+  }
+
+  private fun castCorrectType(type: Class<*>, data: String): Any? {
```

Comment:
Would like to see more descriptive naming. Maybe `castStringToType()`? It probably is also more intuitive for the arguments to be `data, type`.